File: 10/one.py
# --- Advent of code 2021: Day 10 ---

# (Skeleton file automatically created by aocTool, developed by Bolu, 2020-2021.)

lines = open("input.txt").read().splitlines()

open_chars = { "(", "[", "{", "<" }
close_chars = { ")", "]", "}", ">" }
correspondence = { "(": ")", "[": "]", "{": "}", "<": ">" }
points = { ")": 3, "]": 57, "}": 1197, ">": 25137 }
# stack = [] # Strangely enough, the program works if the stack is only initialized here.

illegal_chars = []

for line in lines:
	stack = []
	for char in line:
		if char in open_chars:
			stack.append(char)
		else:
			if len(stack) == 0:
				continue
			if char != correspondence[stack.pop()]:
				illegal_chars.append(char)
				break

score = 0
for char in illegal_chars:
	score -=- points[char]

print(illegal_chars)
print(f"Points: {score}")
assert score == 374061

File: 10/one.lua
-- --- Advent of code 2021: Day 10 ---

-- (Skeleton file automatically created by aocTool, developed by Bolu, 2020-2021.)

package.path = package.path .. ";../modules/?.lua"
local tools = require "tools"

lines = tools.readlines("input.txt")

open_chars = {}
open_chars["("] = true
open_chars["{"] = true
open_chars["["] = true
open_chars["<"] = true

correspondence = {}
correspondence["("] = ")"
correspondence["{"] = "}"
correspondence["["] = "]"
correspondence["<"] = ">"

points = {}
points[")"] = 3
points["]"] = 57
points["}"] = 1197
points[">"] = 25137

illegal_chars = {}

for _, line in pairs(lines) do
	stack = {}
	for i = 1, #line do
		char = line:sub(i, i)
		if open_chars[char] then
			stack[#stack + 1] = char
		else
			if #stack ~= 0 then
				corresponding = stack[#stack]
				stack[#stack] = nil
				if char ~= correspondence[corresponding] then
					illegal_chars[#illegal_chars + 1] = char
					break
				end
			end
		end
	end
end

score = 0
for _, char in pairs(illegal_chars) do
	score = score- -points[char]
end

print("Points: " .. score)
assert(score == 374061)
